equivocation.go

```diff
@@ -0,0 +1,110 @@
+package f3
+
+import (
+	"bytes"
+	"slices"
+	"sync"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// zero value is valid
+type equivocationFilter struct {
+	lk              sync.Mutex
+	currentInstance uint64
+	// sentMessages map unique message slot to its signature
+	sentMessages  map[equivocationKey][]byte
+	activeSenders map[gpbft.ActorID][]peer.ID
+}
+
+type equivocationKey struct {
+	Sender gpbft.ActorID
+	Round  uint64
+	Phase  gpbft.Phase
+}
+
+func (ef *equivocationFilter) formKey(m gpbft.GMessage) equivocationKey {
+	return equivocationKey{
+		Sender: m.Sender,
+		Round:  m.Vote.Round,
+		Phase:  m.Vote.Phase,
+	}
+}
+
+func (ef *equivocationFilter) ProcessBroadcast(peerID peer.ID, m *gpbft.GMessage) bool {
+	ef.lk.Lock()
+	defer ef.lk.Unlock()
+
+	if m.Vote.Instance < ef.currentInstance {
+		// disallow past instances
+		log.Debugw("disallowing broadcast for past message", "sender", m.Sender, "instance",
+			m.Vote.Instance, "currentInstance", ef.currentInstance)
+		return false
+	}
+	// lazy initialization or moved onto new instance
+	if ef.sentMessages == nil || m.Vote.Instance > ef.currentInstance {
+		// bump to next instance
+		ef.currentInstance = m.Vote.Instance
+		ef.sentMessages = make(map[equivocationKey][]byte)
+		ef.activeSenders = make(map[gpbft.ActorID][]peer.ID)
+	}
+
+	key := ef.formKey(*m)
+	sig, ok := ef.sentMessages[key]
+	if ok {
+		log.Debugw("disallowing broadcast due to signature mismatch", "sender", m.Sender,
+			"instance", m.Vote.Instance, "round", m.Vote.Round, "phase", m.Vote.Phase)
+		return bytes.Equal(sig, m.Signature)
+	}
+
+	// save the signature
+	ef.sentMessages[key] = m.Signature
+
+	// save ourselves as one of the senders
+	senders := ef.activeSenders[m.Sender]
+	if !slices.Contains(senders, peerID) {
+		senders = append(senders, peerID)
+		slices.Sort(senders)
+		ef.activeSenders[m.Sender] = senders
+	}
+
+	// if there are multiple senders, only broadcast if we are the smallest one
+	return senders[0] == peerID
```

Comment:
I would log in debug level if this end up being false, along with the cause. Otherwise I suspect it might come across as erroneous behaviour from an operator's perspective (e.g. "Why done my miner not broadcasting F3 messages?")